off_nitrogens/perturb_angle.py

```diff
@@ -0,0 +1,198 @@
+
+def perturb_valence(atom0, atom1, atom2, atom3, theta, verbose=False):
+    """
+    Rotate atom3 in the plane of atom1, atom2, and atom3 by theta degrees.
+    This involves a counterclockwise rotation relative to the normal vector.
+    I.e., if normal vector is (0,0,1) the rotation will be counterclockwise
+    looking from +z onto the xy plane. But if the normal vector is (0,0,-1)
+    the rotation will be clockwise from the same viewpoint.
+
+    Parameters
+    ----------
+    atom0 : numpy array
+        CENTRAL atom coordinates
+    atom1 : numpy array
+        outer atom coordinates
+    atom2 : numpy array
+        outer atom coordinates
+    atom3 : numpy array
+        outer atom coordinates TO BE MOVED
+    theta : float
+        how many degrees by which to rotate
+
+    Returns
+    -------
+    atom* : numpy arrays
+        the coordinates in the same order as given in parameters
+    rot_mat : numpy array
+        three-dimension rotation matrix, returned so that the same
+        matrix can be applied to any atoms attached to atom3.
+
+    """
+
+    # outer atoms are atom1 atom2 atom3. get normal vector to that plane.
+    v1 = atom2-atom1
+    v2 = atom2 - atom3
+    w2 = np.cross(v1, v2)
+    # calculate rotation matrix
+    rot_mat = rotation_matrix(w2, theta)
+    atom3_rot = np.dot(rot_mat, atom3)
+    # print details of geometry
+    if verbose:
+        print("\n>>> Perturbing valence while maintaining improper angle...")
+        # atom being moved is atom3.
+        print("\natom0 original coords:\t",atom0)
+        print("atom1 original coords:\t",atom1)
+        print("atom2 original coords:\t",atom2)
+        print("atom3 original coords:\t",atom3)
+        print("atom3 {:.2f} deg rotated: {}".format(60.,atom3_rot))
+
+        # check improper but make sure the central is first and moved is last
+        print("\nImproper angle, before: ", calc_improper_angle(atom0, atom1, atom2, atom3))
+        print("Improper angle, before: ", calc_improper_angle(atom0, atom2, atom3, atom1))
+        print("Improper angle, before: ", calc_improper_angle(atom0, atom3, atom1, atom2))
+        print("Improper angle, after: ", calc_improper_angle(atom0, atom1, atom2, atom3_rot))
+        print("Improper angle, after: ", calc_improper_angle(atom0, atom2, atom3_rot, atom1))
+        print("Improper angle, after: ", calc_improper_angle(atom0, atom3_rot, atom1, atom2))
+
+        ## check valences
+        print("\nvalence angle 1, before: ", calc_valence_angle(atom0, atom1, atom2))
+        print("valence angle 2, before: ", calc_valence_angle(atom0, atom1, atom3))
+        print("valence angle 3, before: ", calc_valence_angle(atom0, atom2, atom3))
+        print("valence angle 1, after: ", calc_valence_angle(atom0, atom1, atom2))
+        print("valence angle 2, after: ", calc_valence_angle(atom0, atom1, atom3_rot))
+        print("valence angle 3, after: ", calc_valence_angle(atom0, atom2, atom3_rot))
+        print()
```

Comment:
These seem like they might be check if its working print statements, did you want to leave them in for now?

Reply:
I'm not sure what you mean?